### PR TITLE
Create an update slash command

### PR DESCRIPTION
**Pull request by JR-G**
We can now _optionally_ update the values of an incident:

```
/update
```

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -30,4 +30,6 @@
 # Ignore node_modules
 /node_modules
 
-**/.vscode/
\ No newline at end of file
+**/.vscode/
+
+{}
```
Comment (duncanjbrown) [inline .gitignore:35]:
```suggestion
```

File: spec/bot/slash_commands/update_spec.rb [hunk 2]
```diff
@@ -0,0 +1,61 @@
+require 'rails_helper'
+
+describe 'slash_command/update' do
+  include Rack::Test::Methods
+
+  def app
+    SlackRubyBotServer::Api::Middleware.instance
+  end
+
+  context 'without signature checks' do
```
Comment (duncanjbrown) [inline spec/bot/slash_commands/update_spec.rb:10]: Do we need this context block if there's no alternative path?

File: bot/slash_commands/update.rb [hunk 3]
```diff
@@ -0,0 +1,31 @@
+SlackRubyBotServer::Events.configure do |config|
+  config.on :command, '/update' do |command|
+    channel_name = command[:channel_name]
+    channel_id = command[:channel_id]
+    command.logger.info "Updating the incident in #{channel_name}"
```
Comment (duncanjbrown) [inline bot/slash_commands/update.rb:5]: Is this good to log if the channel name might be invalid? Or if we want to know this in all cases, should we log the channel ID and trigger ID too?

File: bot/actions/slack_incident_actions.rb [hunk 4]
```diff
@@ -24,8 +24,59 @@ def open_incident(slack_action)
...
+    )
+
+    threads = []
+    threads << Thread.new { set_channel_details!(channel_id, updated_summary) }
+
+    unless user_invite_list.compact.empty?
+      threads << Thread.new { invite_users!(channel_id, user_invite_list) }
+    end
+
+    unless updated_summary == current_summary
```
Comment (duncanjbrown) [inline bot/actions/slack_incident_actions.rb:49]: Should we make the call to `set_channel_details` depend on this check too?

File: bot/actions/slack_incident_actions.rb [hunk 5]
```diff
@@ -24,8 +24,59 @@ def open_incident(slack_action)
...
+      threads << Thread.new { notify_channel_of_update!(channel_id) }
+    end
+
+    threads.each(&:join)
+  end
+
 private
 
+  def create_member_invite_list(current_members, comms_lead, tech_lead, support_lead)
```
Comment (duncanjbrown) [inline bot/actions/slack_incident_actions.rb:58]: Could `SlackUpdateModal` do this work, given `current_members`, eg `incident.create_invite_list(current_members)`?

File: bot/actions/slack_incident_actions.rb [hunk 6]
```diff
@@ -58,6 +109,11 @@ def notify_channel!(channel, incident_channel_id, title, priority)
                                   text: notification_text)
   end
 
+  def notify_channel_of_update!(channel_id)
```
Comment (duncanjbrown) [inline bot/actions/slack_incident_actions.rb:112]: We could probably at some point come up with a `SlackChannel` object that offered methods like `send_message` etc?

File: bot/actions/slack_incident_listener.rb [hunk 7]
```diff
@@ -4,7 +4,12 @@
 
     begin
       if action[:payload][:view][:app_id] == ENV['SLACK_APP_ID']
-        SlackIncidentActions.new.open_incident(action)
+        if action[:payload][:view][:callback_id] == 'incident-update-modal-identifier'
+          channel_id = Rails.cache.read('channel')
```
Comment (duncanjbrown) [inline bot/actions/slack_incident_listener.rb:8]: I think we established we could make this cache key unique per channel? Not needed for now tho!

File: bot/slash_commands/update.rb [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+        Rails.cache.write('channel', channel_id)
+        { text: 'You’ve updated the incident.' }
+      else
+        { text: 'This is not an incident channel.' }
+      end
+    rescue Slack::Web::Api::Errors::NotInChannel
+      { text: 'This is not an incident channel.' }
+    rescue Slack::Web::Api::Errors::CantInvite
+      { text: 'Cant invite this user' }
+    rescue Slack::Web::Api::Errors::NoUser
```
Comment (duncanjbrown) [inline bot/slash_commands/update.rb:21]: Good to cover these. Should we also be handling errors like this for the incident-create action?

**Review by duncanjbrown — COMMENTED** [review]
🎉  Looking good! Minor comments inline. I'm going to run this before approving 🙃

File: bot/actions/slack_incident_listener.rb (hunk 7 above)
Comment (JR-G) [reply bot/actions/slack_incident_listener.rb:8]: It was originally like [this](https://github.com/DFE-Digital/slack-incident-bot/pull/11/files#diff-41d8fdcddd9bfc947ed7decf57f8a3f04acbe2c48f0f07809b40f6971a09adaeR13)... but I couldn't get it to work on this PR.. so went with the above 🤷‍♂️

File: bot/actions/slack_incident_actions.rb (hunk 6 above)
Comment (JR-G) [reply bot/actions/slack_incident_actions.rb:112]: Will add a new issue to cover this!

File: bot/slash_commands/update.rb (hunk 8 above)
Comment (JR-G) [reply bot/slash_commands/update.rb:21]: Agree – but not too sure on the best errors to cover. We did have an incident recently where the incident description was over the character count.. it errored silently though and still proceeded to create the incident and perform the rest of the actions

Comment (JR-G) [reply bot/slash_commands/update.rb:21]: We do cover some in `bot/actions/slack_incident_listener.rb`

Comment (JR-G) [reply bot/slash_commands/update.rb:21]: have added an extra one

File: bot/actions/slack_incident_actions.rb (hunk 4 above)
Comment (duncanjbrown) [reply bot/actions/slack_incident_actions.rb:46]: :+1: — why not in the same block?

File: bot/slash_commands/update.rb (hunk 3 above)
Comment (duncanjbrown) [reply bot/slash_commands/update.rb:5]: I guess I'm questioning whether there's any value in logging this at all?

Comment (JR-G) [reply bot/slash_commands/update.rb:5]: probably not! – will update

**Review by duncanjbrown — APPROVED** [review]

**Pull request merged**